### PR TITLE
Add support for the sendMessage API endpoint

### PR DESCRIPTION
**Pull request by seanabrahams**

### CODE REVIEW

File: lib/leanplum_api/api.rb [hunk 1]
```diff
@@ -20,6 +20,26 @@ def set_device_attributes(device_attributes, options = {})
...
+    # @param user_id [String] the Leanplum user ID
+    # @param user_ids [Array] Leanplum user IDs (for when you want to send a message to more than one user)
+    # @param device_id [String] the Leanplum device ID
+    # @param create_disposition [String] the policy that determines whether users are created by the API.
+    # @param force [Boolean] whether to send the message regardless of whether the user meets the targeting criteria.
+    # @param values [Hash{Symbol => String, Numeric}] values used to set variables used in the message.
+    # @return [Array<Hash>] the Response(s) from the API
+    def send_message(message_id:, user_id: nil, user_ids: [], device_id: nil, create_disposition: "CreateNever", force: false, values: {}, dev_mode: false)
+      fail ArgumentError, "missing keyword: user_id or user_ids" if user_id.blank? && user_ids.blank?
+      user_ids << user_id if user_id.present?
```
Comment (beingmattlevy) [inline lib/leanplum_api/api.rb]: user_id can never be 0 here. Need to change this to a nil check.

File: spec/api_spec.rb [hunk 2]
```diff
@@ -137,6 +140,96 @@
...
+        end
+        expect(response.size).to eq 1
+        response = response.first
+        expect(response['success']).to eq true
+        expect(response['result']).to eq "Message sent"
+        expect(response['messagesSent']).to eq 1
+        expect(response['error']).to be_blank
+      end
+
+      it "succesfully sends a message to Leanplum's API using device_id" do
```
Comment (beingmattlevy) [inline spec/api_spec.rb:190]: what about an unsuccessful case? like a bad user id or a message id that does not exist?

Comment (seanabrahams) [reply spec/api_spec.rb:190]: @beingmattlevy added!

File: lib/leanplum_api/api.rb [hunk 3]
```diff
@@ -20,6 +20,39 @@ def set_device_attributes(device_attributes, options = {})
...
+    # @param message_id [String] the Leanplum message ID
+    # @param users [Array<Hash{:id => String (required), :device_id => String (optional), :values => Hash (optional)}>] Leanplum users including an optional device_id and vaules hash used to set the variables for the message template
+    # @param create_disposition [String] the policy that determines whether users are created by the API.
+    # @param force [Boolean] whether to send the message regardless of whether the user meets the targeting criteria.
+    # @param values [Hash{Symbol => String, Numeric}] values used to set variables used in the message.
+    # @return [Array<Hash>] the Response(s) from the API
+    def send_messages(message_id:, users:, create_disposition: "CreateNever", force: false, values: {}, dev_mode: false)
+      fail ArgumentError, "Cannot accept more than 50 users" if users.size > 50
+      validation_results = users_valid?(users)
+      fail ArgumentError, "users failed validation: #{validation_results[1]}" if validation_results != true
```
Comment (beingmattlevy) [inline lib/leanplum_api/api.rb]: I don't get what your doing here. The users_valid? returns a boolean, no? Not sure what your doing with the Array in validation_results. I see some array magic going on in the method, but its not clear if you are creating a polymorphic return or a signal from the method as to which user has an issue. Why not just raise from the validation method?

File: lib/leanplum_api/api.rb [hunk 4]
```diff
@@ -266,5 +299,26 @@ def build_event_attributes_hash(event_hash, options = {})
...
+        userId: user_id,
+        deviceId: device_id,
+        createDisposition: create_disposition,
+        force: force,
+        values: values.to_json,
+        devMode: dev_mode
+      }
+    end
+
+    def users_valid?(users)
```
Comment (beingmattlevy) [inline lib/leanplum_api/api.rb]: This is interesting logic. Lets write a test for this or at least add a comment for this method describing the error messaging paradigm.

**Review by beingmattlevy — CHANGES_REQUESTED** [review]

File: lib/leanplum_api/api.rb (hunk 3 above)
Comment (seanabrahams) [reply lib/leanplum_api/api.rb]: Yeah, it's non-obvious. The `users_valid?` predicate method returns `true` or `false, error_message` which is not a common practice in ruby-land. I'll re-think this.

File: lib/leanplum_api/api.rb (hunk 4 above)
Comment (seanabrahams) [reply lib/leanplum_api/api.rb]: Tests exist for this. I've re-worked how this works and will push an update.

File: lib/leanplum_api/api.rb [hunk 5]
```diff
@@ -20,6 +20,38 @@ def set_device_attributes(device_attributes, options = {})
...
+    # POSTs multiple messages to Leanplum's sendMessage API endpoint
+    #
+    # @param message_id [String] the Leanplum message ID
+    # @param users [Array<Hash{:id => String (required), :device_id => String (optional), :values => Hash (optional)}>] Leanplum users including an optional device_id and vaules hash used to set the variables for the message template
+    # @param create_disposition [String] the policy that determines whether users are created by the API.
+    # @param force [Boolean] whether to send the message regardless of whether the user meets the targeting criteria.
+    # @param values [Hash{Symbol => String, Numeric}] values used to set variables used in the message.
+    # @return [Array<Hash>] the Response(s) from the API
+    def send_messages(message_id:, users:, create_disposition: "CreateNever", force: false, values: {}, dev_mode: false)
+      fail ArgumentError, "Cannot accept more than 50 users due to Leanplum API restrictions" if users.size > 50
```
Comment (beingmattlevy) [inline lib/leanplum_api/api.rb]: move this inside the validate_users
Also need a present? check before you call a size check.

**Review by beingmattlevy — APPROVED** [review]

**Pull request merged**